src/v/storage/disk_log_impl.cc

```diff
@@ -500,48 +500,55 @@ ss::future<> disk_log_impl::adjacent_merge_compact(
 
 segment_set disk_log_impl::find_sliding_range(
   const compaction_config& cfg, std::optional<model::offset> new_start_offset) {
+    if (
+      _last_compaction_window_start_offset
+      <= _segs.front()->offsets().get_base_offset()) {
+        // If this evaluates to true, it is likely because local retention has
+        // removed segments. e.g, segments ([0],[1],[2]), have been garbage
+        // collected to segments ([2]), while _last_window_compaction_offset
+        // == 1. To avoid compaction getting stuck in this situation, we reset
+        // the compaction window offset here.
+        vlog(
+          gclog.debug,
+          "[{}] start offset ({}) <= base offset of front segment ({}), "
+          "resetting compaction window start offset",
+          config().ntp(),
+          _last_compaction_window_start_offset,
+          _segs.front()->offsets().get_base_offset());
+
+        _last_compaction_window_start_offset.reset();
+    }
+
     // Collect all segments that have stable data.
     segment_set::underlying_t buf;
     for (const auto& seg : _segs) {
         if (seg->has_appender() || !seg->has_compactible_offsets(cfg)) {
             // Stop once we get to an unstable segment.
             break;
         }
+        if (
+           _last_compaction_window_start_offset.has_value()
+           && (seg->offsets().get_base_offset()
+                >= _last_compaction_window_start_offset.value())) {
+            // Force clean segment production by compacting down to the
+            // start of the log before considering new segments in the
+            // compaction window.
+            break;
+        }
         if (
           new_start_offset
           && seg->offsets().get_base_offset() < new_start_offset.value()) {
             // Skip over segments that are being truncated.
             continue;
         }
+
         buf.emplace_back(seg);
     }
     segment_set segs(std::move(buf));
     if (segs.empty()) {
         return segs;
     }
 
-    // If a previous sliding window compaction ran, and there are no new
-    // segments, segments at the start of that window and above have been
-    // fully deduplicated and don't need to be compacted further.
-    //
-    // If there are new segments that have not been compacted, we can't make
-    // this claim, and compact everything again.
-    if (
```

Comment:
It may look a bit scary to remove this chunk of code, but this is functionally performing the same "clean-compacted" check we now perform [here](https://github.com/redpanda-data/redpanda/pull/23380/commits/4045b9f90f9be8a814c3bb15941f5b15bb690b64#diff-8caa0c9719de25bd209fa674959d9327cd55237de2fb4683761acf9bcc6de217R594).

We want to leave these segments in the compaction range so that they can be self-compacted in the case that they contain tombstones, and then filter them out afterwards before proceeding to sliding window compaction.